core/src/main/java/com/graphhopper/routing/ch/SplitArray2D.java

```diff
@@ -122,4 +122,17 @@ void clear(int n) {
         mids[n] = 0;
     }
 
+    private void grow(int n) {
+        // todo: think about grow factor: trimming and then doubling the size might be not what we want, rather
+        // increase by 50% or something?
```

Comment:
How large are these sub-arrays on average? Could it make sense to initialize them with the average+20% then they do not need to grow that often and the size mismatch could be lower?

Reply:
Ggood question. In this commit: https://github.com/graphhopper/graphhopper/pull/2132/commits/1c2ca0f9c421ff36b645c31fc2d66852717911cd I trimmed the arrays after the prepare graph is built for the first time to save some memory at the beginning of the contraction. For maximum speed it would of course be better to initialize them with some expected size (but no idea so far how important this is). I am currently running the import with this trimming to compare to the version without the trimming.

Reply:
I did a quick check for Bayern (~6mio shortcuts) and Germany (~28mio shortcuts) and the array sizes seem to be relatively stable against changes of the map size. Before the contraction the average size of the arrays (number of in+out edges) is around 4.5 (makes sense: 2.25 in/out edges per node) and average mids[n] is almost exactly half of sizes[n] (on average the number of in edges is the same as out edges -> makes sense). 

The maximum size during the contraction is about 100 and the maximum mid is around 50 for both maps. 

So I think its already safe to say that we can store mid/size in a single int to save some (not a lot) memory (around 900MB for planet). For the resizing strategy I think it probably does not really matter, because the arrays never get very large and maybe its even the best approach to trim them as I do currently to make sure memory is minimized.

In my opinion if we change anything here we should try linking the edge objects to each other and remove the sub-arrays entirely, but I would postpone this until we feel like its definitely needed (and we are sure it improves anything).

Reply:
> to trim them as I do currently to make sure memory is minimized.

Yes, this sounds reasonable